File: internal/ratelimit/ratelimit.go
// Package ratelimit implements a per-key sliding-window rate limiter.
package ratelimit

import (
	"sync"
	"time"
)

// Limiter allows at most max events per window per key. A non-positive max
// disables the limit (every call is allowed). It is safe for concurrent use.
type Limiter struct {
	max    int
	window time.Duration

	mu     sync.Mutex
	events map[int64][]time.Time
	now    func() time.Time // injectable clock for tests
}

// New builds a Limiter for max events per window.
func New(max int, window time.Duration) *Limiter {
	return &Limiter{
		max:    max,
		window: window,
		events: make(map[int64][]time.Time),
		now:    time.Now,
	}
}

func (l *Limiter) prune(key int64, now time.Time) []time.Time {
	events := l.events[key]
	cutoff := now.Add(-l.window)
	i := 0
	for i < len(events) && events[i].Before(cutoff) {
		i++
	}
	if i > 0 {
		events = events[i:]
	}
	l.events[key] = events
	return events
}

// Allow records and permits an event if the key is under quota.
func (l *Limiter) Allow(key int64) bool {
	if l.max <= 0 {
		return true
	}
	l.mu.Lock()
	defer l.mu.Unlock()
	now := l.now()
	events := l.prune(key, now)
	if len(events) >= l.max {
		return false
	}
	l.events[key] = append(events, now)
	return true
}

// RetryAfter reports how long until a slot frees up (0 if one is available now).
func (l *Limiter) RetryAfter(key int64) time.Duration {
	if l.max <= 0 {
		return 0
	}
	l.mu.Lock()
	defer l.mu.Unlock()
	now := l.now()
	events := l.prune(key, now)
	if len(events) < l.max {
		return 0
	}
	d := l.window - now.Sub(events[0])
	if d < 0 {
		return 0
	}
	return d
}

File: legacy/dropzone47/ratelimit.py
import time
from collections import defaultdict, deque
from typing import Deque, Dict


class RateLimiter:
    """Fixed-quota sliding-window limiter keyed by an arbitrary id (e.g. user id).

    Allows at most ``max_events`` calls per ``window_seconds``. A non-positive
    ``max_events`` disables the limit (every call is allowed).
    """

    def __init__(self, max_events: int, window_seconds: int) -> None:
        self.max_events = max_events
        self.window = window_seconds
        self._events: Dict[int, Deque[float]] = defaultdict(deque)

    def _now(self) -> float:
        return time.time()

    def allow(self, key: int) -> bool:
        if self.max_events <= 0:
            return True
        now = self._now()
        events = self._events[key]
        while events and now - events[0] > self.window:
            events.popleft()
        if len(events) >= self.max_events:
            return False
        events.append(now)
        return True

    def retry_after(self, key: int) -> int:
        """Seconds until the oldest event in the window expires (0 if a slot is free)."""
        if self.max_events <= 0:
            return 0
        events = self._events.get(key)
        if not events or len(events) < self.max_events:
            return 0
        return max(0, int(self.window - (self._now() - events[0])))
